src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -194,18 +199,22 @@ class LedgerImpl(vm: VM, blockchain: BlockchainImpl, blockchainConfig: Blockchai
 
     val refundGasFn = pay(stx.senderAddress, (totalGasToRefund * gasPrice).toUInt256) _
     val payMinerForGasFn = pay(Address(blockHeader.beneficiary), (executionGasToPayToMiner * gasPrice).toUInt256) _
+
+    val worldAfterPayments = (refundGasFn andThen payMinerForGasFn)(resultWithErrorHandling.world)
+
     val deleteAccountsFn = deleteAccounts(resultWithErrorHandling.addressesToDelete) _
+    val deleteTouchedAccountsFn = deleteTouchedAccounts _
     val persistStateFn = InMemoryWorldStateProxy.persistState _
 
-    val world2 = (refundGasFn andThen payMinerForGasFn andThen deleteAccountsFn andThen persistStateFn)(resultWithErrorHandling.world)
+    val world2 = (deleteAccountsFn andThen deleteTouchedAccountsFn andThen persistStateFn)(worldAfterPayments)
```

Comment:
a random thought: do you think it would make sense to always have a list of touched accounts in WorldStateProxy (and not consider noEmptyAccounts at all there), and instead have a single condition here, that would delete them (if noEmptyAccounts=true or else do nothing)?

Reply:
that way we wouldn't have to maintain the `Option` logic there, and we wouldn't need the additional flag. Of course I might miss something obvious here :)